ztoc/ztoc_builder.go

```diff
@@ -17,238 +17,86 @@
-	if err != nil {
-		return nil, err
-	}
-
-	fs, err := getFileSize(gzipFile)
-	if err != nil {
-		return nil, err
-	}
-
-	digests, err := getPerSpanDigests(gzipFile, int64(fs), index)
-	if err != nil {
-		return nil, err
-	}
-
-	checkpoints, err := index.Bytes()
-	if err != nil {
-		return nil, err
-	}
+type buildOption struct {
+	algorithm string
+}
 
-	toc := TOC{
-		Metadata: fm,
-	}
+// BuildOption specifies a change to `buildOption` when building a ztoc.
+type BuildOption func(opt *buildOption) error
 
-	compressionInfo := CompressionInfo{
-		MaxSpanID:   index.MaxSpanID(),
-		SpanDigests: digests,
-		Checkpoints: checkpoints,
+// WithCompression specifies which compression algorithm is used by the layer.
+func WithCompression(algorithm string) BuildOption {
+	return func(opt *buildOption) error {
+		opt.algorithm = algorithm
+		return nil
 	}
-
-	return &Ztoc{
-		Version:                 "0.9",
-		TOC:                     toc,
-		CompressedArchiveSize:   fs,
-		UncompressedArchiveSize: uncompressedArchiveSize,
-		BuildToolIdentifier:     buildToolIdentifier,
-		CompressionInfo:         compressionInfo,
-	}, nil
 }
 
-func getPerSpanDigests(gzipFile string, fileSize int64, index *compression.GzipZinfo) ([]digest.Digest, error) {
-	file, err := os.Open(gzipFile)
-	if err != nil {
-		return nil, fmt.Errorf("could not open file for reading: %w", err)
-	}
-	defer file.Close()
-
-	var digests []digest.Digest
-	var i compression.SpanID
-	maxSpanID := index.MaxSpanID()
-	for i = 0; i <= maxSpanID; i++ {
-		var (
-			startOffset = index.SpanIDToCompressedOffset(i)
-			endOffset   compression.Offset
-		)
-
-		if index.HasBits(i) {
-			startOffset--
-		}
-
-		if i == maxSpanID {
-			endOffset = compression.Offset(fileSize)
-		} else {
-			endOffset = index.SpanIDToCompressedOffset(i + 1)
-		}
-
-		section := io.NewSectionReader(file, int64(startOffset), int64(endOffset-startOffset))
-		dgst, err := digest.FromReader(section)
-		if err != nil {
-			return nil, fmt.Errorf("unable to compute digest for section; start=%d, end=%d, file=%s, size=%d", startOffset, endOffset, gzipFile, fileSize)
-		}
-		digests = append(digests, dgst)
-	}
-	return digests, nil
+// Builder holds a single `TocBuilder` that builds toc, and one `ZinfoBuilder`
+// *per* compression algorithm that builds zinfo. `TocBuilder` is shared by different
+// compression algorithms. Which `ZinfoBuilder` is used depends on the compression
+// algorithm used by the layer.
+type Builder struct {
+	tocBuilder    TocBuilder
+	zinfoBuilders map[string]ZinfoBuilder
 }
 
-func getGzipFileMetadata(gzipFile string, index *compression.GzipZinfo) ([]FileMetadata, compression.Offset, error) {
-	file, err := os.Open(gzipFile)
-	if err != nil {
-		return nil, 0, fmt.Errorf("could not open file for reading: %v", err)
+// NewBuilder creates a `Builder` used to build ztocs.
+func NewBuilder() *Builder {
+	return &Builder{
+		tocBuilder: NewTocBuilder(),
+		zinfoBuilders: map[string]ZinfoBuilder{
```

Comment:
Instead of having a map here, can we pass in the `ZinfoBuilder` as an argument to `NewBuilder`?

Reply:
We don't know beforehand which compression will be used in `IndexBuilder.Build(ctx, img)`. `img` might have layers with both `gzip` and `zstd`, since `compression` is a [per-layer property](https://github.com/awslabs/soci-snapshotter/blob/12b4ceff9a9c82fd588a42a25c3965bc8ef9c270/soci/soci_index.go#L294).

Using a map makes `ztoc.Builder` management simpler since the `ztoc.Builder` will just be part of `IndexBuilder` (1 index builder has 1 ztoc builder which handles layers with different compressions). Also, I don't think there will be many compressions (IIRC, containerd recognize gzip and zstd and others will be just unknow).

If `ZinfoBuilder` is passed as an argument, that means:

1. a `ztoc.Builder` is bind to that `ZinfoBuilder` (e.g., gzip). So the `IndexBuilder` needs to figure out, for each layer, which compression is used and pass that to `NewBuilder` 
2. `IndexBuilder` need to call `NewBuilder` for every layer.
3. `ztoc.Builder` has a `tocBuilder` that supports different compression but a `zinfoBuilder` that supports only one compression.

Reply:
SGTM.

For `tocBuilder` to support a new algorithm, only a helper func that _creates a tar reader from the compression reader_ is needed and the rest logic is the same.

For `zinfoBuilder`, different algorithms require different implementation (e.g., `GzipZinfoBuilder` and underlying `compression.GzipZinfo`).

I'll add a register to `ztoc.Builder` that accepts a compression name, a tar reader helper, a `zinfoBuilder` instance, and we can have a default register for gzip (e.g., `ztoc.RegisterGzip`).

Reply:
Added a register func to `ztoc.Builder`. By default `NewBuilder` registers `gzip`, and users can call `builder.RegisterCompressionAlgorithm` to support/register new compressions.

Initially I implmeneted it as an `Option` to `NewBuilder` (i.e., `WithCompressionAlgorithm` which is just a wraper of `RegisterCompressionAlgorithm`) but didn't see much benefit and it's not some config/option that has to be finalized during init. So I only kept `RegisterCompressionAlgorithm`.